lib.network: first/last and next addresses in a network

woojiq opened the pull request:
### Description of changes

Depends on #318712 and #326765

* Add functions to find the first, last and next addresses in a network using
bitwise operations.
* Write unit tests for such functions.

Mentor: @Janik-Haag 

### Things done
- [x] Tested, as applicable:
* [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
* and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
* or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).
- [x] Formatted using nixfmt-rfc-style

Add a 👍 [reaction](https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to [pull requests you find important](https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

Janik-Haag commented on lib/network/default.nix:
```diff
@@ -0,0 +1,65 @@
+
+      - [addr] An IPv6 address with optional prefix length.
+    */
+    fromString =
```
Instead of modifying `fromString`, we should have something like `lib.ipv6.firstAddress` and `lib.ipv6.lastAddress`.

woojiq replied:
Hmm, can you explain roughly how you want to use the library? From my point of view, I expect something like `fromString <addr>` and then the user takes whatever data they need from the returned attr set.

Janik-Haag replied:
I guess that's fair, esp. since if we implement it correctly the values are lazily evaluated, which means it doesn't really increase the cost. But I still think we should expose the function to end users because it's more ergonomic to call `lib.ipv6.lastAddress "f00::c0ffee/64"` vs `(lib.ipv6.fromString "f00::c0ffee/64").lastAddress`. There will also be functions in the future that aren't really incorporatable into `fromString`, for example imagine something like `lib.ipv6.arithmetic "f00::c0ffee/64" (-50)`

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/showcase-nixpkgs-network-library/48897/1

woojiq committed:
lib.network: add ipv6Addr type, mkNetworkingOption option

Add ipv6 ipv6Addr and ipv6AddrAttrs types.
Add ipv6.isValidIpStr function to be used as the ipv6Addr type checker.
Add mkNetworkingOption that uses `coercedTo` function to convert string
IP to parsed attr set ipv6AddrAttrs.

Change the design of `ipv6.fromString` function. It used to be created
to return an attrset of data like `nextAddr`, `firstAddr`, etc. But
now it returns an attrset with a string representation of the address
(default and Cidr) and contains an internal representation. That way,
other future functions like `nextAddr` can easily take such type and
create a new one, because there is no need to reparse the string.

woojiq committed:
lib.network: add network library README.md

README.md explains the design and purpose of the network library. At the
time of writing, IPv4 functionality is not yet merged.

woojiq committed:
lib.network: ipv6 first and last addresses in a subnet

Add functions to calculate first and last addresses.
Add unit tests.

woojiq committed:
lib.network: ipv6 next address

Add a functions to calculate the next address from the specified or null.
Add unit tests.
